core/src/main/java/org/elasticsearch/index/query/IndicesQueryBuilder.java

```diff
@@ -81,4 +90,53 @@ protected void doXContent(XContentBuilder builder, Params params) throws IOExcep
     public String getWriteableName() {
         return NAME;
     }
+
+    @Override
+    protected Query doToQuery(QueryShardContext context) throws IOException {
+        if (context.matchesIndices(indices)) {
+            return innerQuery.toQuery(context);
+        }
+        return noMatchQuery.toQuery(context);
+    }
+    
+    @Override
+    public QueryValidationException validate() {
+        QueryValidationException validationException = null;
+        if (this.innerQuery == null) {
+            validationException = addValidationError("inner query cannot be null", validationException);
+        }
+        if (this.indices == null || this.indices.length == 0) {
+            validationException = addValidationError("list of indices cannot be null or empty", validationException);
+        }
+        validationException = validateInnerQuery(innerQuery, validationException);
+        validationException = validateInnerQuery(noMatchQuery, validationException);
+        return validationException;
+    }
+
+    @Override
+    protected IndicesQueryBuilder doReadFrom(StreamInput in) throws IOException {
+        IndicesQueryBuilder indicesQueryBuilder = new IndicesQueryBuilder(in.readQuery(), in.readStringArray());
+        indicesQueryBuilder.noMatchQuery = in.readQuery();
+        return indicesQueryBuilder;
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeQuery(innerQuery);
+        out.writeStringArray(indices);
+        out.writeQuery(noMatchQuery);
+    }
+    
+    @Override
+    public int doHashCode() {
+        int result = Objects.hash(innerQuery, noMatchQuery);
+        return 31 * result + Arrays.hashCode(indices);
```

Comment:
Is this because of https://github.com/elastic/elasticsearch/issues/12145? Just curious.